### PR TITLE
fix: reading multiple holding registers in modbus input plugin

### PR DESCRIPTION
**Pull request by garciaolais**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

- Fix https://github.com/influxdata/telegraf/issues/7227

if you try to read multiple consecutive addresses , modbus return error "quantity 346 must be between 1 and 125".
this happens because we have a limitation with the number of registers that we can retrieve, 125 for holding registers and Input registers and 2000 for discrete inputs and coils

### CODE REVIEW

**Review by Jagularr — COMMENTED** [review]
Logic looks good to me

**Comment by helenosheaa** [comment]
Looks good @garciaolais thanks for opening up another PR!

File: plugins/inputs/modbus/modbus.go [hunk 1]
```diff
@@ -217,19 +217,29 @@ func (m *Modbus) InitRegister(fields []fieldContainer, name string) error {
 	sort.Slice(addrs, func(i, j int) bool { return addrs[i] < addrs[j] })
 
 	ii := 0
+	maxQuantity := 1
 	var registersRange []registerRange
+	if name == cDiscreteInputs || name == cCoils {
+		maxQuantity = 2000
+	}
+
+	if name == cInputRegisters || name == cHoldingRegisters {
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus.go:226]: Can we make this an `else if` please.

File: plugins/inputs/modbus/modbus.go [hunk 2]
```diff
@@ -217,19 +217,29 @@ func (m *Modbus) InitRegister(fields []fieldContainer, name string) error {
...
+		maxQuantity = 125
+	}
 
 	// Get range of consecutive integers
 	// [1, 2, 3, 5, 6, 10, 11, 12, 14]
 	// (1, 3) , (5, 2) , (10, 3), (14 , 1)
 	for range addrs {
+		quantity := 1
 		if ii < len(addrs) {
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus.go:235]: Not directly related to the patch, but wouldn't it make sense to rather do
```suggestion
		if ii >= len(addrs) {
			break
		}
```

**Review by srebhan — COMMENTED** [review]
Looks good basically, but I'd prefer to have the checking for max grouped registers in an else-if to make it obvious that it is exclusive.

File: plugins/inputs/modbus/modbus.go (hunk 2 above)
Comment (garciaolais) [reply plugins/inputs/modbus/modbus.go:233]: I've tested but that change broke some tests

**Review by srebhan — APPROVED** [review]
Perfect thanks!

**Review by helenosheaa — APPROVED** [review]

File: plugins/inputs/modbus/modbus.go (hunk 2 above)
Comment (srebhan) [reply plugins/inputs/modbus/modbus.go:233]: Very strange as from a logic point-of-view this should be equivalent. This is not related to the PR anyway, so consider this as resolved.

Comment (garciaolais) [reply plugins/inputs/modbus/modbus.go:233]: Let me double check that

Comment (garciaolais) [reply plugins/inputs/modbus/modbus.go]: @srebhan  you are right, your suggestion works 👍

File: plugins/inputs/modbus/modbus.go [hunk 3]
```diff
@@ -217,23 +217,32 @@ func (m *Modbus) InitRegister(fields []fieldContainer, name string) error {
...
-				ii++
-			}
+		for ii < len(addrs)-1 && addrs[ii+1]-addrs[ii] == 1 && quantity < maxQuantity {
+			end = addrs[ii+1]
 			ii++
-			registersRange = append(registersRange, registerRange{start, end - start + 1})
+			quantity++
 		}
+		ii++
+		registersRange = append(registersRange, registerRange{start, end - start + 1})
```
Comment (srebhan) [inline plugins/inputs/modbus/modbus.go:245]: Hmmm I'm pretty sure that the additional `i++` here means we should change `end-start+1` to `end-start`!? However, I would feel much more safe if we remove the additional `i++` and keep the range calculation as is... :-/

**Review by srebhan — COMMENTED** [review]
I think the register-range changed. Could you please check my comment in the code!?

**Comment by srebhan** [comment]
@garciaolais can you please comment on my question above regarding the new increment of `end`!?

**Comment by garciaolais** [comment]
@srebhan sorry for the delay but I really doesn't have free time to check this. Let me chance to retake it and I hope to test it this week.

File: plugins/inputs/modbus/modbus.go (hunk 3 above)
Comment (garciaolais) [reply plugins/inputs/modbus/modbus.go]: we can remove `i++` but the other change _`end-start+1 `to `end-start`_  brake some test

**Comment by srebhan** [comment]
@garciaolais sure, just wanted to make sure it is not forgotten. Please let me know about the results of your tests!

File: plugins/inputs/modbus/modbus.go (hunk 3 above)
Comment (garciaolais) [reply plugins/inputs/modbus/modbus.go]: Hi @srebhan 

Do you have any other comments about this issue?

**Review by srebhan — APPROVED** [review]
Looks good to me assuming that you tested the code @garciaolais ... :-)

**Review by ssoroka — CHANGES_REQUESTED** [review]
This all looks good. One ask is that the tests don't validate the generated metric. It's possible that it runs without errors but doesn't give you the response you'd expect. 
Does it make sense to add some lines to at least one of the tests to make sure the metric was generated as expected?

**Review by ssoroka — APPROVED** [review]

**Review by ssoroka — CHANGES_REQUESTED** [review]
Actually, would still be great to get a test showing the expected metrics are generated properly.

**Review by srebhan — APPROVED** [review]
Looks good to me.

**Review by ssoroka — APPROVED** [review]

**Pull request merged**